examples/src/test/scala/JedisExample.scala

```diff
@@ -58,83 +60,129 @@ object DataSources {
-          case (b, o) =>
-            Sync[F].pure({
-              val res = o.readObject()
-              Option(res.asInstanceOf[A])
-            })
-        })({
-          case (b, o) =>
-            Sync[F].delay({
-              b.close()
-              o.close()
-            })
-        })
-      }
-    }
-
+  def deserialize[F[_], A](bin: ByteArray)(
+      implicit S: Sync[F]
+  ): F[Option[A]] = {
+    type Resources = (ObjectInputStream, ByteArrayInputStream)
+
+    def resources[F[_]](implicit S: Sync[F]): Resource[F, Resources] =
+      for {
+        byte <- byteInputStream(bin)
+        in   <- inputStream(byte)
+      } yield (in, byte)
+
+    resources.use({
+      case (in, byte) =>
+        S.delay {
+          val obj = in.readObject()
+          Try(obj.asInstanceOf[A]).toOption
+        }
+    })
+  }
 }
 
 case class RedisCache[F[_]: Sync](host: String) extends DataCache[F] {
-  private val binaryClient = new BinaryJedis(host)
+  private val pool = new JedisPool(host)
 
-  def client: F[BinaryJedis] =
-    Sync[F].pure(binaryClient)
+  def connection: Resource[F, Jedis] =
+    Resource.fromAutoCloseable(Sync[F].delay(pool.getResource))
 
   private def get(i: Array[Byte]): F[Option[Array[Byte]]] =
-    Sync[F].bracket(client)({ c =>
-      Sync[F].pure(Option(c.get(i)))
-    })({ in =>
-      Sync[F].pure(in.close())
+    connection.use({ c =>
+      Sync[F].delay(Option(c.get(i)))
     })
 
   private def set(i: Array[Byte], v: Array[Byte]): F[Unit] =
-    Sync[F].bracket(client)({ c =>
-      Sync[F].pure(c.set(i, v)).void
-    })({ in =>
-      Sync[F].pure(in.close())
+    connection.use({ c =>
+      Sync[F].delay(c.set(i, v)).void
     })
 
+  private def bulkSet(ivs: List[(Array[Byte], Array[Byte])]): F[Unit] =
+    connection.use(c =>
+      Sync[F].delay({
+        val pipe = c.pipelined
+        ivs.foreach(i => pipe.set(i._1, i._2))
+        pipe.sync
+      }))
+
   private def identity[I, A](i: I, data: Data[I, A]): Array[Byte] =
     Binary.fromString(s"${data.identity} ${i}")
 
-  def lookup[I, A](i: I, data: Data[I, A]): F[Option[A]] =
-    get(identity(i, data)) >>= { case (raw) => Binary.deserialize(raw) }
+  override def lookup[I, A](i: I, data: Data[I, A]): F[Option[A]] =
+    for {
+      raw <- get(identity(i, data))
+      result <- raw match {
+        case None    => Sync[F].pure(None)
+        case Some(r) => Binary.deserialize[F, A](r)
+      }
+    } yield result
 
-  def insert[I, A](i: I, v: A, data: Data[I, A]): F[DataCache[F]] =
+  override def insert[I, A](i: I, v: A, data: Data[I, A]): F[DataCache[F]] =
     for {
       s <- Binary.serialize(v)
       _ <- set(identity(i, data), s)
     } yield this
+
+  override def bulkInsert[I, A](vs: List[(I, A)], data: Data[I, A])(
+      implicit M: Monad[F]
+  ): F[DataCache[F]] =
+    for {
+      bin <- vs.traverse({
+        case (id, v) =>
+          for {
+            bin <- Binary.serialize(v)
+          } yield (identity(id, data), bin)
```

Comment:
There is an `identity` function in the scala standard library. Perhaps the function could be renamed to avoid confussion.